src/wasm/wasm-type.cpp

```diff
@@ -772,15 +827,45 @@ struct SignatureTypeCache {
 static SignatureTypeCache nominalSignatureCache;
 
 // Keep track of the constructed recursion groups.
-static std::mutex recGroupsMutex;
-static std::vector<std::unique_ptr<RecGroupInfo>> recGroups;
+struct RecGroupStore {
+  std::mutex mutex;
+  std::unordered_set<RecGroupStructure> canonicalGroups;
+  std::vector<std::unique_ptr<RecGroupInfo>> builtGroups;
```

Comment:
I'm not sure how to read this. What are "canonical groups" and what are "built groups"?

Reply:
I'll add comments. `canonicalGroups` is used to do the actual hashing and deduplication of groups, including singleton groups that do not have associated `RecGroupInfo`s, while `builtGroups` exists just to hold the unique pointers keeping the `RecGroupInfo`s alive.